File: js/projections/client/wait.ts
import { DataListWait, DataWait } from "@fraym/proto/dist/index.freym.projections.delivery";
import { Filter, getProtobufDataFilter } from "./filter";

export interface Wait {
    timeout?: number;
    conditionFilter: Filter;
}

export const getProtobufDataWait = (wait?: Wait): DataWait | undefined => {
    if (!wait) {
        return undefined;
    }

    return {
        conditionFilter: getProtobufDataFilter(wait.conditionFilter),
        timeout: (wait.timeout ?? 0).toString(),
    };
};

export interface ListWait {
    timeout?: number;
    condition: DataListWaitCondition[];
}

export interface DataListWaitCondition {
    operation: string;
    value: number;
}

export const getProtobufDataListWait = (wait?: ListWait): DataListWait | undefined => {
    if (!wait) {
        return undefined;
    }

    return {
        condition: wait.condition.map(condition => ({
            operation: condition.operation,
            value: condition.value.toString(),
        })),
        timeout: (wait.timeout ?? 0).toString(),
    };
};

File: go/projections/wait.go
package projections

import (
	"time"

	"github.com/fraym/freym-api/go/proto/projections/deliverypb"
	"github.com/samber/lo"
)

type Wait struct {
	ConditionFilter  *Filter
	ConditionTimeout time.Duration
}

func (w *Wait) toDeliveryWait() *deliverypb.DataWait {
	if w == nil || w.ConditionFilter == nil {
		return nil
	}

	return deliverypb.DataWait_builder{
		ConditionFilter: w.ConditionFilter.toProtobufFilter(),
		Timeout:         w.ConditionTimeout.Milliseconds(),
	}.Build()
}

type JsonWait struct {
	ConditionFilter  *JsonFilter
	ConditionTimeout time.Duration
}

func (w *JsonWait) toDeliveryWait() *deliverypb.DataWait {
	if w == nil || w.ConditionFilter == nil {
		return nil
	}

	return deliverypb.DataWait_builder{
		ConditionFilter: w.ConditionFilter.toProtobufFilter(),
		Timeout:         w.ConditionTimeout.Milliseconds(),
	}.Build()
}

type ListWait struct {
	Condition        []ListWaitCondition
	ConditionTimeout time.Duration
}

type ListWaitCondition struct {
	Operation string
	Value     int64
}

func (w *ListWait) toDeliveryListWait() *deliverypb.DataListWait {
	if w == nil || w.Condition == nil {
		return nil
	}

	return deliverypb.DataListWait_builder{
		Condition: lo.Map(w.Condition, func(condition ListWaitCondition, _ int) *deliverypb.DataListWaitCondition {
			return deliverypb.DataListWaitCondition_builder{
				Operation: condition.Operation,
				Value:     condition.Value,
			}.Build()
		}),
		Timeout: w.ConditionTimeout.Milliseconds(),
	}.Build()
}
